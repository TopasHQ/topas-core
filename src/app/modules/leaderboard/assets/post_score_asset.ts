import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from 'lisk-sdk';

import { Highscore, LeaderboardModuleAccountProps, LeaderboardModuleChainData, ModuleId } from '../../../types';
import {
    createDateTime,
    createHighscoreEssentials,
    createTopasAppEssentials,
    createUserEssentials,
} from '../../../utils/helpers';
import { getTopasAppById, getTopasUserData } from '../../../utils/reducer_handlers';
import { getStateStoreData } from '../../../utils/store';
import { validateHexString, validateIsPublished } from '../../../utils/validation';
import { LEADERBOARD_ASSET_IDS } from '../constants';
import { LEADERBOARD_KEY, leaderboardModuleSchema } from '../schemas';

type Props = {
	score: number;
	appId: string;
};

export class PostScoreAsset extends BaseAsset {
	public name = 'postScore';
	public id = LEADERBOARD_ASSET_IDS.postScore;

	public schema = {
		$id: 'leaderboard/postScore-asset',
		title: 'PostScoreAsset transaction asset for leaderboard module',
		type: 'object',
		required: ['appId', 'score'],
		properties: {
			appId: {
				dataType: 'string',
				fieldNumber: 1,
			},
			score: {
				dataType: 'uint32',
				fieldNumber: 2,
			},
		},
	};

	public validate({ asset }: ValidateAssetContext<Props>): void {
		validateHexString(asset.appId);
	}

	public async apply({ asset, transaction, stateStore, reducerHandler }: ApplyAssetContext<Props>): Promise<void> {
		const account = await stateStore.account.getOrDefault<LeaderboardModuleAccountProps>(transaction.senderAddress);
		const topasUser = await getTopasUserData(reducerHandler, { address: account.address, errorOnEmpty: true });
		const topasApp = await getTopasAppById(reducerHandler, { id: asset.appId });
		validateIsPublished(topasApp);

		const stateStoreData = await getStateStoreData<LeaderboardModuleChainData>(stateStore, ModuleId.Leaderboard);

		const highscore: Highscore = {
			user: createUserEssentials(account, topasUser),
			app: createTopasAppEssentials(topasApp),
			score: asset.score,
			createdAt: createDateTime(),
		};

		const highscoreEssentials = createHighscoreEssentials(highscore);

		// Check if user has already posted highscore for app
		const moduleIndex = stateStoreData.highscores.findIndex(
			score => score.user.username === topasUser.username && score.app.id === asset.appId,
		);

		// If so, check if score is higher and update both module and account data
		if (moduleIndex !== -1) {
			if (asset.score <= stateStoreData.highscores[moduleIndex].score) {
				throw new Error(`Score is equal or less than existing score by user.`);
			}

			stateStoreData.highscores[moduleIndex] = highscore;

			const accountHighscoreIndex = account.leaderboard.highscores.findIndex(score => score.appId === asset.appId);
			account.leaderboard.highscores[accountHighscoreIndex] = highscoreEssentials;
		}

		// If not, add score to both module and account data
		if (moduleIndex === -1) {
			account.leaderboard.highscores.push(highscoreEssentials);
			stateStoreData.highscores.push(highscore);
		}

		await stateStore.account.set(account.address, account);
		await stateStore.chain.set(LEADERBOARD_KEY, codec.encode(leaderboardModuleSchema, stateStoreData));
	}
}
