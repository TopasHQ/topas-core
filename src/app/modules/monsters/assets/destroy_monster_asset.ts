import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from 'lisk-sdk';

import { ModuleId, MonstersModuleChainData } from '../../../types';
import { getStateStoreData } from '../../../utils/store';
import { validateHexString, validateRegistration } from '../../../utils/validation';
import { MONSTERS_ASSET_IDS } from '../constants';
import { MONSTERS_KEY, monstersModuleSchema } from '../schemas';

type Props = {
	id: string;
};

export class DestroyMonsterAsset extends BaseAsset {
	public name = 'destroyMonster';
	public id = MONSTERS_ASSET_IDS.destroyMonster;

	public schema = {
		$id: 'monsters/destroyMonster-asset',
		title: 'DestroyMonsterAsset transaction asset for monsters module',
		type: 'object',
		required: ['id'],
		properties: {
			id: {
				dataType: 'string',
				fieldNumber: 1,
			},
		},
	};
	public validate({ asset }: ValidateAssetContext<Props>): void {
		validateHexString(asset.id);
	}

	public async apply({ asset, transaction, stateStore, reducerHandler }: ApplyAssetContext<Props>): Promise<void> {
		await validateRegistration(reducerHandler, transaction.senderAddress);
		const stateStoreData = await getStateStoreData<MonstersModuleChainData>(stateStore, ModuleId.Monsters);

		const monster = stateStoreData.activeMonsters.find(m => m.id === asset.id);

		if (!monster) {
			throw Error('Monster does not exist.');
		}

		stateStoreData.activeMonsters = stateStoreData.activeMonsters.filter(m => m.id !== asset.id);

		await reducerHandler.invoke('token:credit', {
			address: transaction.senderAddress,
			amount: monster.reward,
		});

		await stateStore.chain.set(MONSTERS_KEY, codec.encode(monstersModuleSchema, stateStoreData));
	}
}
