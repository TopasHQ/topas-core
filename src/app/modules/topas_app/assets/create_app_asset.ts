import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from 'lisk-sdk';

import config from '../../../config';
import { ModuleId } from '../../../types';
import { bufferToHex, createMeta, createTopasAppEssentials, createUserEssentials } from '../../../utils/helpers';
import { getTopasUserData } from '../../../utils/reducer_handlers';
import { getStateStoreData } from '../../../utils/store';
import { validateEntranceFee, validateTransactionFee } from '../../../utils/validation';
import { TOPAS_APP_ASSET_IDS, TOPAS_APP_FEES, TOPAS_APP_MODULE_KEY } from '../constants';
import { topasAppModuleSchema } from '../schemas';
import { CreateAppAssetProps, TopasApp, TopasAppModuleAccountProps, TopasAppModuleChainData } from '../types';

export class CreateAppAsset extends BaseAsset {
	public name = 'createApp';
	public id = TOPAS_APP_ASSET_IDS.createApp;
	public fee = TOPAS_APP_FEES.createApp;

	public schema = {
		$id: 'topasApp/createApp-asset',
		title: 'CreateAppAsset transaction asset for topasApp module',
		type: 'object',
		required: ['type', 'title', 'description', 'tipsEnabled', 'entranceFee'],
		properties: {
			type: {
				dataType: 'uint32',
				fieldNumber: 1,
			},
			title: {
				dataType: 'string',
				fieldNumber: 2,
				minLength: config.appTitleMinLength,
				maxLength: config.appTitleMaxLength,
			},
			description: {
				dataType: 'string',
				fieldNumber: 3,
				minLength: config.appDescriptionMinLength,
				maxLength: config.appDescriptionMaxLength,
			},
			tipsEnabled: {
				fieldNumber: 4,
				dataType: 'boolean',
			},
			entranceFee: {
				fieldNumber: 5,
				dataType: 'uint64',
			},
		},
	};

	public validate({ transaction, asset }: ValidateAssetContext<CreateAppAssetProps>): void {
		validateTransactionFee(transaction, this.fee);
		validateEntranceFee(asset.entranceFee);
	}

	public async apply({
		asset,
		transaction,
		stateStore,
		reducerHandler,
	}: ApplyAssetContext<CreateAppAssetProps>): Promise<void> {
		const account = await stateStore.account.getOrDefault<TopasAppModuleAccountProps>(transaction.senderAddress);
		const topasUser = await getTopasUserData(reducerHandler, { address: account.address, errorOnEmpty: true });
		const stateStoreData = await getStateStoreData<TopasAppModuleChainData>(stateStore, ModuleId.TopasApp);

		const appExists = !!stateStoreData.apps.find(p => p.data.title === asset.title);
		if (appExists) {
			throw new Error(`App already exists.`);
		}

		const app: TopasApp = {
			meta: createMeta(),
			data: {
				...asset,
				id: bufferToHex(transaction.id),
				creator: createUserEssentials(account, topasUser),
				isPublished: false,
				numOfUses: 0,
			},
		};

		stateStoreData.apps.push(app);
		account.topasApp.appsCreated.push(createTopasAppEssentials(app));

		await stateStore.chain.set(TOPAS_APP_MODULE_KEY, codec.encode(topasAppModuleSchema, stateStoreData));
		await stateStore.account.set(account.address, account);
	}
}
