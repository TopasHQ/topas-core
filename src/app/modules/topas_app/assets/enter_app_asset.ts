import { ApplyAssetContext, BaseAsset, codec } from 'lisk-sdk';

import { TICKER } from '../../../constants';
import { ModuleId, TopasAppModuleAccountProps, TopasAppModuleChainData } from '../../../types';
import { beddowsToLsk, createTopasAppEssentials, senderIsAppCreator, senderOwnsApp } from '../../../utils/helpers';
import { getStateStoreData, getTopasApp } from '../../../utils/store';
import { validateIsPublished, validateRegistration } from '../../../utils/validation';
import { TOPAS_APP_ASSET_IDS } from '../constants';
import { TOPAS_APP_KEY, topasAppModuleSchema } from '../schemas';

type Props = {
	appId: string;
};

export class EnterAppAsset extends BaseAsset {
	public name = 'enterApp';
	public id = TOPAS_APP_ASSET_IDS.enterApp;

	public schema = {
		$id: 'topasApp/enterApp-asset',
		title: 'EnterAppAsset transaction asset for topasApp module',
		type: 'object',
		required: ['appId'],
		properties: {
			appId: {
				dataType: 'string',
				fieldNumber: 1,
			},
		},
	};

	public async apply({ asset, transaction, stateStore, reducerHandler }: ApplyAssetContext<Props>): Promise<void> {
		const account = await stateStore.account.getOrDefault<TopasAppModuleAccountProps>(transaction.senderAddress);
		await validateRegistration(reducerHandler, account.address);

		const stateStoreData = await getStateStoreData<TopasAppModuleChainData>(stateStore, ModuleId.TopasApp);

		const app = getTopasApp(stateStoreData, asset.appId);
		validateIsPublished(app);

		const isCreator = senderIsAppCreator(app, transaction);
		if (!isCreator) {
			if (senderOwnsApp(app, account)) {
				throw new Error(`Sender already owns app.`);
			}

			const { entranceFee } = app.data;
			const totalCost = entranceFee + transaction.fee;

			const accountBalance: bigint = await reducerHandler.invoke('token:getBalance', {
				address: transaction.senderAddress,
			});

			if (totalCost > accountBalance) {
				throw new Error(`Sender balance is too low. Required ${beddowsToLsk(totalCost)} ${TICKER}.`);
			}

			await reducerHandler.invoke('token:debit', {
				address: transaction.senderAddress,
				amount: entranceFee,
			});

			await reducerHandler.invoke('token:credit', {
				address: app.data.creator.address,
				amount: entranceFee,
			});

			app.data.numOfUses += 1;
			account.topasApp.appsPurchased.push(createTopasAppEssentials(app));
		}

		await stateStore.chain.set(TOPAS_APP_KEY, codec.encode(topasAppModuleSchema, stateStoreData));
		await stateStore.account.set(account.address, account);
	}
}
