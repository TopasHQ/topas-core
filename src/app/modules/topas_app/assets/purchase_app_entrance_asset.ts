import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from 'lisk-sdk';

import { TICKER } from '../../../constants';
import { ModuleId, TopasAppPurchase } from '../../../types';
import {
    beddowsToLsk,
    bufferToHex,
    createDateTime,
    createTopasAppEssentials,
    senderIsAppCreator,
} from '../../../utils/helpers';
import { getStateStoreData, getTopasApp } from '../../../utils/store';
import { validateHexString, validateIsPublished, validateRegistration } from '../../../utils/validation';
import { TOPAS_APP_MODULE_KEY } from '../constants';
import { purchaseAppEntranceAssetPropsSchema, topasAppModuleSchema } from '../schemas';
import { PurchaseAppEntranceAssetProps, TopasAppMode, TopasAppModuleAccountProps, TopasAppModuleChainData } from '../types';

export class PurchaseAppEntranceAsset extends BaseAsset {
	public name = 'purchaseAppEntrance';
	public id = 4;
	public schema = purchaseAppEntranceAssetPropsSchema;

	public validate({ asset }: ValidateAssetContext<PurchaseAppEntranceAssetProps>): void {
		validateHexString(asset.appId);
	}

	public async apply({
		asset,
		transaction,
		stateStore,
		reducerHandler,
	}: ApplyAssetContext<PurchaseAppEntranceAssetProps>): Promise<void> {
		const account = await stateStore.account.getOrDefault<TopasAppModuleAccountProps>(transaction.senderAddress);
		await validateRegistration(reducerHandler, account.address);

		const stateStoreData = await getStateStoreData<TopasAppModuleChainData>(stateStore, ModuleId.TopasApp);

		const app = getTopasApp(stateStoreData, asset.appId);
		validateIsPublished(app);

		if (senderIsAppCreator(app, transaction)) {
			throw new Error('Sender is creator of app.');
		}

		const { entranceFee } = app.data;
		const totalCost = entranceFee + transaction.fee;

		const accountBalance: bigint = await reducerHandler.invoke('token:getBalance', {
			address: transaction.senderAddress,
		});

		if (totalCost > accountBalance) {
			throw new Error(`Sender balance is too low. Required ${beddowsToLsk(totalCost)} ${TICKER}.`);
		}

		await reducerHandler.invoke('token:debit', {
			address: transaction.senderAddress,
			amount: entranceFee,
		});

		if (app.data.mode === TopasAppMode.feeToChest) {
			app.data.prize += entranceFee;
		}

		if (app.data.mode !== TopasAppMode.feeToChest) {
			await reducerHandler.invoke('token:credit', {
				address: app.data.creator.address,
				amount: entranceFee,
			});
		}

		app.data.numOfUses += 1;
		await stateStore.chain.set(TOPAS_APP_MODULE_KEY, codec.encode(topasAppModuleSchema, stateStoreData));

		const appPurchase: TopasAppPurchase = {
			...createTopasAppEssentials(app),
			purchaseId: bufferToHex(transaction.id),
			createdAt: createDateTime(),
		};
		account.topasApp.appsPurchases.push(appPurchase);
		await stateStore.account.set(account.address, account);
	}
}
