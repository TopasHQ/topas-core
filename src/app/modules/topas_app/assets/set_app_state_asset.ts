import { ApplyAssetContext, BaseAsset, codec, cryptography } from 'lisk-sdk';

import config from '../../../config';
import { ModuleId, TopasAppModuleChainData } from '../../../types';
import { updateMeta } from '../../../utils/helpers';
import { getStateStoreData, getTopasApp } from '../../../utils/store';
import { TOPAS_APP_ASSET_IDS } from '../constants';
import { TOPAS_APP_KEY, topasAppModuleSchema } from '../schemas';

type Props = {
	appId: string;
	isPublished: boolean;
};

export class SetAppStateAsset extends BaseAsset {
	public name = 'setAppState';
	public id = TOPAS_APP_ASSET_IDS.setAppState;

	public schema = {
		$id: 'topasApp/setAppState-asset',
		title: 'SetAppStateAsset transaction asset for topasApp module',
		type: 'object',
		required: ['appId', 'isPublished'],
		properties: {
			appId: {
				dataType: 'string',
				fieldNumber: 1,
			},
			isPublished: {
				fieldNumber: 2,
				dataType: 'boolean',
			},
		},
	};

	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<Props>): Promise<void> {
		const stateStoreData = await getStateStoreData<TopasAppModuleChainData>(stateStore, ModuleId.TopasApp);
		const app = getTopasApp(stateStoreData, asset.appId);

		if (app.data.isPublished === asset.isPublished) {
			throw new Error(`App's isPublished state is already set to ${String(asset.isPublished)}.`);
		}

		// TODO: Research and fix potential security risk
		const isAdmin = config.adminAccounts.includes(cryptography.bufferToHex(transaction.senderAddress));
		if (!isAdmin) {
			throw new Error(`Sender is not privileged to set app state.`);
		}

		app.meta = updateMeta(app.meta);
		app.data = {
			...app.data,
			isPublished: asset.isPublished,
		};

		await stateStore.chain.set(TOPAS_APP_KEY, codec.encode(topasAppModuleSchema, stateStoreData));
	}
}
