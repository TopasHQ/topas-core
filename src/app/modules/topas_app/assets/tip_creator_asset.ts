import { ApplyAssetContext, BaseAsset, ValidateAssetContext } from 'lisk-sdk';

import { TICKER } from '../../../constants';
import { ModuleId, TopasAppModuleAccountProps, TopasAppModuleChainData } from '../../../types';
import { beddowsToLsk, senderIsAppCreator } from '../../../utils/helpers';
import { getStateStoreData, getTopasApp } from '../../../utils/store';
import { validateHexString, validateIsPublished, validateRegistration, validateTipAmount } from '../../../utils/validation';
import { TOPAS_APP_ASSET_IDS } from '../constants';

type Props = {
	appId: string;
	tipAmount: bigint;
};

export class TipCreatorAsset extends BaseAsset {
	public name = 'tipCreator';
	public id = TOPAS_APP_ASSET_IDS.tipCreator;

	public schema = {
		$id: 'topasApp/tipCreator-asset',
		title: 'TipCreatorAsset transaction asset for topasApp module',
		type: 'object',
		required: ['appId', 'tipAmount'],
		properties: {
			appId: {
				dataType: 'string',
				fieldNumber: 1,
			},
			tipAmount: {
				dataType: 'uint64',
				fieldNumber: 2,
			},
		},
	};

	public validate({ asset }: ValidateAssetContext<Props>): void {
		validateTipAmount(asset.tipAmount);
		validateHexString(asset.appId);
	}

	public async apply({ asset, transaction, stateStore, reducerHandler }: ApplyAssetContext<Props>): Promise<void> {
		const account = await stateStore.account.getOrDefault<TopasAppModuleAccountProps>(transaction.senderAddress);
		await validateRegistration(reducerHandler, account.address);

		const stateStoreData = await getStateStoreData<TopasAppModuleChainData>(stateStore, ModuleId.TopasApp);

		const app = getTopasApp(stateStoreData, asset.appId);
		validateIsPublished(app);

		const isCreator = senderIsAppCreator(app, transaction);
		if (isCreator) {
			throw new Error(`Sender is creator of app.`);
		}

		const { tipAmount } = asset;
		const totalCost = tipAmount + transaction.fee;

		const accountBalance: bigint = await reducerHandler.invoke('token:getBalance', {
			address: transaction.senderAddress,
		});

		if (totalCost > accountBalance) {
			throw new Error(`Sender balance is too low. Required ${beddowsToLsk(totalCost)} ${TICKER}.`);
		}

		await reducerHandler.invoke('token:debit', {
			address: transaction.senderAddress,
			amount: tipAmount,
		});

		await reducerHandler.invoke('token:credit', {
			address: app.data.creator.address,
			amount: tipAmount,
		});

		await stateStore.account.set(account.address, account);
	}
}
