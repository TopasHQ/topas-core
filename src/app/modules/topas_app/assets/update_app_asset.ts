import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from 'lisk-sdk';

import config from '../../../config';
import { ModuleId, TopasAppModuleAccountProps, TopasAppModuleChainData } from '../../../types';
import { senderIsAppCreator, updateMeta } from '../../../utils/helpers';
import { getStateStoreData, getTopasApp } from '../../../utils/store';
import { validateEntranceFee, validateHexString, validateTransactionFee } from '../../../utils/validation';
import { TOPAS_APP_ASSET_IDS, TOPAS_APP_FEES } from '../constants';
import { TOPAS_APP_KEY, topasAppModuleSchema } from '../schemas';

type Props = {
	id: string;
	description: string;
	tipsEnabled: boolean;
	entranceFee: bigint;
};

export class UpdateAppAsset extends BaseAsset {
	public name = 'updateApp';
	public id = TOPAS_APP_ASSET_IDS.updateApp;
	public fee = TOPAS_APP_FEES.updateApp;

	public schema = {
		$id: 'topasApp/updateApp-asset',
		title: 'UpdateAppAsset transaction asset for topasApp module',
		type: 'object',
		required: ['id', 'description', 'tipsEnabled', 'entranceFee'],
		properties: {
			id: {
				dataType: 'string',
				fieldNumber: 1,
			},
			description: {
				dataType: 'string',
				fieldNumber: 2,
				minLength: config.appDescriptionMinLength,
				maxLength: config.appDescriptionMaxLength,
			},
			tipsEnabled: {
				fieldNumber: 3,
				dataType: 'boolean',
			},
			entranceFee: {
				fieldNumber: 4,
				dataType: 'uint64',
			},
		},
	};

	public validate({ transaction, asset }: ValidateAssetContext<Props>): void {
		validateTransactionFee(transaction, this.fee);
		validateEntranceFee(asset.entranceFee);
		validateHexString(asset.id);
	}

	public async apply({ asset, transaction, stateStore }: ApplyAssetContext<Props>): Promise<void> {
		const account = await stateStore.account.getOrDefault<TopasAppModuleAccountProps>(transaction.senderAddress);
		const stateStoreData = await getStateStoreData<TopasAppModuleChainData>(stateStore, ModuleId.TopasApp);

		const app = getTopasApp(stateStoreData, asset.id);

		const isCreator = senderIsAppCreator(app, transaction);
		if (!isCreator) {
			throw new Error(`Sender is not creator of app.`);
		}

		app.meta = updateMeta(app.meta);
		app.data = {
			...app.data,
			...asset,
		};

		await stateStore.chain.set(TOPAS_APP_KEY, codec.encode(topasAppModuleSchema, stateStoreData));
		await stateStore.account.set(account.address, account);
	}
}
