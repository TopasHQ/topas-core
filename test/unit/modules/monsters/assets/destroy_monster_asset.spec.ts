import { DestroyMonsterAsset } from '../../../../../src/app/modules/monsters/assets/destroy_monster_asset';

describe('DestroyMonsterAsset', () => {
	let transactionAsset: DestroyMonsterAsset;

	beforeEach(() => {
		transactionAsset = new DestroyMonsterAsset();
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(1);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('destroyMonster');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		describe('schema validation', () => {
			it.todo('should throw errors for invalid schema');
			it.todo('should be ok for valid schema');
		});
	});

	describe('apply', () => {
		describe('valid cases', () => {
			it.todo('should update the state store');
		});

		describe('invalid cases', () => {
			it.todo('should throw error');
		});
	});
});
