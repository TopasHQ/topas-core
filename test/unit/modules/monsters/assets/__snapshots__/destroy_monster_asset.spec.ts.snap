// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestroyMonsterAsset constructor should have valid schema 1`] = `
Object {
  "$id": "monsters/destroyMonster-asset",
  "properties": Object {
    "id": Object {
      "dataType": "string",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "id",
  ],
  "title": "DestroyMonsterAsset transaction asset for monsters module",
  "type": "object",
}
`;
